### PR TITLE
Optimized schema building process without Python `_WalkCoreSchema`

### PR DESCRIPTION
**Pull request by MarkusSintonen**
## Change Summary

PR removes expensive core schema walking with `_WalkCoreSchema` which is replaced by `pydantic-core` schema traversal. The traversal is a one pass operation when the necessary information is gathered from the schema tree for cleaning it up. Inlining of "definitions-ref"s is done in place for the relevant nodes. This avoids rewalking the schema and copy operations. Similarly discriminated unions are handled by directly substituting the appropriate schema node. PR also removes completely unnecessary extra schema walking with invalid generics handling.

Used big model file to run benchmarks https://gist.github.com/Viicos/bd9b3b0e3fbee62838626b07b311f3af.
PR reduces the load time of the big model file (on my MB Pro M3 Max):
Old load time: 3.44s
New load time: 2.65s

Below CodSpeed results look very promising.

Requires `pydantic-core` [PR](https://github.com/pydantic/pydantic-core/pull/1487) which this PR points into.

## Related issue number

Relates to https://github.com/pydantic/pydantic/issues/6768 Need later verification how much this PR helps in real usage. Also it might not help with all the perf issues elsewhere in schema generation process.

## Checklist

* [x] The pull request title is a good summary of the changes - it will be used in the changelog
* [x] Unit tests for the changes exist
* [x] Tests pass on CI
* [x] Documentation reflects the changes where applicable
* [x] My PR is ready to review, **please add a comment including the phrase "please review" to assign reviewers**

Selected Reviewer: @sydney-runkle

### CODE REVIEW

**Comment by MarkusSintonen** [comment]
please review

**Comment by sydney-runkle** [comment]
@MarkusSintonen,

So awesome. We're wrapping up v2.10 this week, then will be pivoting to a performance focus for v2.11. For that reason, we might be a bit delayed in pushing this PR across the line, but I look forward to thoroughly reviewing + helping to get these perf improvements on `main`.

I'll certainly give it a review this week, just wanted to indicate that we shouldn't expect a speedy merge here.

**Comment by MarkusSintonen** [comment]
> I'll certainly give it a review this week, just wanted to indicate that we shouldn't expect a speedy merge here.

No worries, thanks!

**Comment by sydney-runkle** [comment]
Releasing v2.10.0b1 tomorrow, then pivoting to perf stuff :)

**Comment by sydney-runkle** [comment]
> Releasing v2.10.0b1 tomorrow, then pivoting to perf stuff :)

Well, clearly not. Shooting for by the end of the week on this now, we're held up by a new exciting feature + a fix in integrating with `pydantic-core`.

**Comment by sydney-runkle** [comment]
@MarkusSintonen,

What made you decide to close https://github.com/pydantic/pydantic/pull/10622 and reintroduce the schema walking?

**Comment by sydney-runkle** [comment]
I see https://github.com/pydantic/pydantic/pull/10622#issuecomment-2422064776

Could you clarify more regarding the following:

> Closed PR tracked python instance references tracking during schema building. Which could be a huge breaking change for users due to it disallowing schema node copying when building schemas.

We're looking to implement more caching on the schema build side, so wanted to check regarding compatibility.

**Comment by sydney-runkle** [comment]
A few ideas, up front:
* We should go ahead and move `CoreMetadata` to `pydantic-core` - I can work on this 👍 
* We should explicitly add the discriminator keys to `CoreMetadata` to make things easier
* We should be able to remove `InvalidSchema` from `pydantic-core`, it seems?

**Comment by MarkusSintonen** [comment]
> What made you decide to close https://github.com/pydantic/pydantic/pull/10622 and reintroduce the schema walking? I'm assuming we need that info for inlining and tagged discriminator updates?

That approach wasn't safe as it required keeping track of the schema node instances accurately. But those assumption blow up badly when users construct custom core schemas. Eg via `__get_pydantic_core_schema__`. Also the schemas are very mutable so that also breaks assumption about being able to track the node instances accurately.

It is much safer to keep the walking behaviour to gather the needed info (but doing it more optimally, ie the pydantic-core side walking has really minimal overhead).

**Comment by MarkusSintonen** [comment]
> We should be able to remove InvalidSchema from pydantic-core, it seems?

Yes, good catch! Its now removed in both PRs

**Comment by MarkusSintonen** [comment]
> > We should be able to remove InvalidSchema from pydantic-core, it seems?
> 
> Yes, good catch! Its now removed in both PRs

Oh yeah, `pydantic-core` cleanup needs to happen later. As there are the integration tests that wont pass otherwise because it requires this PR to be merged first.

File: pydantic/_internal/_generate_schema.py [hunk 1]
```diff
@@ -2425,12 +2392,108 @@ def _common_field(
...
 
 
+class CollectedInvalid(Exception):
+    pass
+
+
+_NOT_INLINABLE_META_KEYS: set[str] = {
+    *CoreMetadata.__annotations__.keys(),
+    _discriminated_union.CORE_SCHEMA_METADATA_DISCRIMINATOR_PLACEHOLDER_KEY,
+    TAGGED_UNION_TAG_KEY,
```
[the remark this answers is not in the document]
Comment (sydney-runkle) [reply pydantic/_internal/_generate_schema.py:2402]: That being said, we should probably have some sort of check that after cleaning, those keys are no longer present in the metadata, as (if I recall), we pop them on application.

File: pydantic/_internal/_generate_schema.py [hunk 2]
```diff
@@ -2425,12 +2392,108 @@ def _common_field(
...
+        # Def schema instances of another model, not to be exposed via get_def
+        self._unpacked_definitions: dict[str, CoreSchema] = {}
+
+    def get_def(self, ref: str) -> CoreSchema | None:
+        return self._definitions.get(ref)
+
+    def reference_schema(self, schema: CoreSchema) -> core_schema.DefinitionReferenceSchema:
+        ref = schema['ref']  # type: ignore
+        self._definitions[ref] = schema
+        return core_schema.definition_reference_schema(ref)
```
Comment (sydney-runkle) [inline pydantic/_internal/_generate_schema.py:2422]: Wouldn't it be better to just pass `ref` here as well, as we only call this when we know `ref` is present and have confirmed that? Removes the need for the `# type: ignore`

Comment (sydney-runkle) [reply pydantic/_internal/_generate_schema.py:2422]: I don't love this new function. This pattern happens a lot, so I'm not opposed to coupling the last two lines, but I think we need to name it something different, because the name only implies the logic of the last line

File: pydantic/_internal/_generate_schema.py [hunk 3]
```diff
@@ -2425,12 +2392,108 @@ def _common_field(
...
-        self.seen: set[str] = set()
-        self.definitions: dict[str, core_schema.CoreSchema] = {}
+        self.used_deferred_discriminators = False
+        self._recursively_seen: set[str] = set()
+        self._definitions: dict[str, CoreSchema] = {}
+        # Def schema instances of another model, not to be exposed via get_def
+        self._unpacked_definitions: dict[str, CoreSchema] = {}
+
+    def get_def(self, ref: str) -> CoreSchema | None:
+        return self._definitions.get(ref)
```
Comment (sydney-runkle) [inline pydantic/_internal/_generate_schema.py:2417]: Is this adding much value?

File: pydantic/_internal/_generate_schema.py [hunk 4]
```diff
@@ -533,36 +530,21 @@ def _apply_discriminator_to_union(
...
-        schema = self.collect_definitions(schema)
-        schema = simplify_schema_references(schema)
-        if collect_invalid_schemas(schema):
-            raise self.CollectedInvalid()
-        schema = _discriminated_union.apply_discriminators(schema)
+    def clean_schema(self, schema: CoreSchema, *, deep_copy: bool) -> CoreSchema:
+        """Cleans the schema contents by substituting definition-refs and discriminators. Given schema must be the
+        one just given by generate_schema. Does in place substituting to the schema by mutating it for performance
+        reasons. Use deep_copy to avoid mutating the original schema.
+        """
```
Comment (sydney-runkle) [inline pydantic/_internal/_generate_schema.py:542]: Let's be even more explicit here.

Can we:
* Explain subsituting definition refs
* Explain why / how we add tagged info to discriminators?

This isn't your doing at all, but our internal `GenerateSchema` logic is poorly documented, so I think it's good to have thorough improvements as a standard for PRs here going forward.

File: pydantic/_internal/_generate_schema.py [hunk 5]
```diff
@@ -1655,8 +1626,8 @@ def _sequence_schema(self, items_type: Any) -> core_schema.CoreSchema:
 
         item_type_schema = self.generate_schema(items_type)
         list_schema = core_schema.list_schema(item_type_schema)
+        json_schema = core_schema.list_schema(self.generate_schema(items_type))
```
Comment (sydney-runkle) [inline pydantic/_internal/_generate_schema.py:1629]: Surprised we don't need the copy here anymore

**Review by sydney-runkle — COMMENTED** [review]
Still got more to review, but here's a start!

Admittedly, a bit hard to follow the defs / refs resolution logic, so will take another pass at this soon. Going to review the `pydantic-core` PR first.

File: pydantic/_internal/_known_annotated_metadata.py [hunk 6]
```diff
@@ -187,7 +187,6 @@ def apply_known_metadata(annotation: Any, schema: CoreSchema) -> CoreSchema | No
 
     from ._validators import NUMERIC_VALIDATOR_LOOKUP, forbid_inf_nan_check
 
-    schema = schema.copy()
```
[the remark this answers is not in the document]
Comment (MarkusSintonen) [reply pydantic/_internal/_known_annotated_metadata.py:190]: `apply_known_metadata` is called in single spot where the schema is already `copy`ed (also previously) so it did double copying. As its very internal function it seemed like it did the copying for no good reason.

File: tests/test_generics.py [hunk 7]
```diff
@@ -1794,17 +1791,16 @@ def module():
...
 
         class M2(BaseModel, Generic[V3]):
-            m: Union[M1[int, V3], V3]
+            m1: Union[M1[int, V3], V3]
 
         M1.model_rebuild()
 
     M1 = module.M1
-
-    # assert M1.__pydantic_core_schema__ == {}
```
[the remark this answers is not in the document]
Comment (MarkusSintonen) [reply tests/test_generics.py:1806]: [schema.txt](https://github.com/user-attachments/files/17805150/schema.txt)

The line was commented out. Someone had forgotten some debug stuff here. But there is the full core schema of `M1` (not parametrized)

File: tests/test_generics.py [hunk 8]
```diff
@@ -1844,7 +1840,7 @@ class B1(BaseModel, Generic[S1]):
             a1: 'B2[S1]'
 
         class B2(BaseModel, Generic[S2]):
-            a2: 'B1[S2]'
+            a2: Optional['B1[S2]'] = None
```
[the remark this answers is not in the document]
Comment (MarkusSintonen) [reply tests/test_generics.py:1843]: I just wanted to get the model possibly to test. It was impossible to initialize previously as B2 and B1 depend on each other and both have a single required field. That chain is not possible to break. But well I can change it back so the test doesnt do anything else than just seeing it wont crash.

File: tests/test_generics.py [hunk 9]
```diff
@@ -2239,35 +2234,6 @@ class GenericPydanticModel(BaseModel, Generic[T]):
...
-    try:
-        with generic_recursion_self_type(Model, (int,)):
-            # Make sure that something has been added to the contextvar-managed recursive types cache
-            assert recursively_defined_type_refs()
-            raise TestingException
-    except TestingException:
-        pass
-
-    # Make sure that an exception causes the contextvar-managed recursive types cache to be reset
-    assert not recursively_defined_type_refs()
```
[the remark this answers is not in the document]
Comment (MarkusSintonen) [reply tests/test_generics.py:2268]: `recursively_defined_type_refs` doesnt exist anymore which the test tested (same as here https://github.com/pydantic/pydantic/pull/10655#discussion_r1847088052)

File: tests/test_type_alias_type.py [hunk 10]
```diff
@@ -400,8 +400,8 @@ class MyModel(BaseModel):
     assert MyModel(my_int_seq=range(1, 4)).my_int_seq == [1, 2, 3]
 
     assert MyModel.model_json_schema() == {
-        '$defs': {'MySeq_int_': {'items': {'type': 'integer'}, 'type': 'array'}},
-        'properties': {'my_int_seq': {'$ref': '#/$defs/MySeq_int_'}},
+        '$defs': {'MyIntSeq': {'items': {'type': 'integer'}, 'type': 'array'}},
+        'properties': {'my_int_seq': {'$ref': '#/$defs/MyIntSeq'}},
```
[the remark this answers is not in the document]
Comment (MarkusSintonen) [reply tests/test_type_alias_type.py:404]: No I dont think so, it should just make the json schema names more clear for TypeAliasType so it uses the user defined names. Comes from the "proper" usage of `GenerateSchema` `_Definitions`

File: pydantic/_internal/_generics.py [hunk 11]
```diff
@@ -432,14 +432,6 @@ def generic_recursion_self_type(
...
 
 
-def recursively_defined_type_refs() -> set[str]:
-    visited = _generic_recursion_cache.get()
-    if not visited:
-        return set()  # not in a generic recursion, so there are no types
-
-    return visited.copy()  # don't allow modifications
-
-
```
[the remark this answers is not in the document]
Comment (MarkusSintonen) [reply pydantic/_internal/_generics.py:442]: Its the same thing as here https://github.com/pydantic/pydantic/pull/10655#discussion_r1847088052

So the generics definition handling was quite a bit wonky previously. But it feels getting a bit more sane after the walking refactoring.

File: pydantic/_internal/_generate_schema.py (hunk 2 above)
Comment (MarkusSintonen) [reply pydantic/_internal/_generate_schema.py:2420]: `create_reference_to_schema`? 🤔 

> Wouldn't it be better to just pass ref here as well

Well maybe, but I would still have an assertion to check that there wont be coming anything corrupt when ref is coming from arg. So: `assert schema['ref'] == ref`

File: pydantic/_internal/_generate_schema.py (hunk 3 above)
Comment (MarkusSintonen) [reply pydantic/_internal/_generate_schema.py:2415]: So adding `__getitem__` instead?

Comment (MarkusSintonen) [reply pydantic/_internal/_generate_schema.py:2415]: I wouldnt make it public to not cause a mayhem of things mutating the registry

File: pydantic/_internal/_generate_schema.py (hunk 5 above)
Comment (MarkusSintonen) [reply pydantic/_internal/_generate_schema.py:1634]: It was deep copying before here, but instead just regenerate a new instance for the "json schema" side. So we dont need the deep copying traversing the whole object tree. Its more consistent with the rest of the logic then.

File: pydantic/_internal/_discriminated_union.py [hunk 12]
```diff
@@ -217,7 +188,7 @@ def _apply_to_root(self, schema: core_schema.CoreSchema) -> core_schema.CoreSche
             # was flattened by pydantic_core.
             # However, it still may make sense to apply the discriminator to this schema,
             # as a way to get discriminated-union-style error messages, so we allow this here.
-            schema = core_schema.union_schema([schema])
+            schema = core_schema.union_schema([schema.copy()])
```
[the remark this answers is not in the document]
Comment (MarkusSintonen) [reply pydantic/_internal/_discriminated_union.py]: Cleaned it up. Removed this change, instead for simplicity do the shallow copy in caller side in `finalize_schema_defs_inplace`. It might do some unnecessary copying in above cases `if schema['type'] == 'nullable'` or `if schema['type'] == 'definitions'` but they feel quite rare.

File: pydantic/_internal/_schema_generation_shared.py [hunk 13]
```diff
@@ -115,12 +114,26 @@ def resolve_ref_schema(self, maybe_ref_schema: core_schema.CoreSchema) -> core_s
...
             return self.resolve_ref_schema(maybe_ref_schema['schema'])
         return maybe_ref_schema
+
+
+def get_existing_core_schema(obj: Any) -> core_schema.CoreSchema | None:
+    if (
+        hasattr(obj, '__dict__')
+        # In some cases (e.g. a stdlib dataclass subclassing a Pydantic dataclass),
+        # doing an attribute access to get the schema will result in the parent schema
+        # being fetched. Thus, only look for the current obj's dict:
```
[the remark this answers is not in the document]
Comment (MarkusSintonen) [reply pydantic/_internal/_schema_generation_shared.py]: Done, used the other previous comment

File: pydantic/_internal/_discriminated_union.py [hunk 14]
```diff
@@ -28,34 +25,8 @@ def __init__(self, ref: str) -> None:
...
-
-        metadata = s.get('metadata', {})
-        discriminator = metadata.pop(CORE_SCHEMA_METADATA_DISCRIMINATOR_PLACEHOLDER_KEY, None)
-        if discriminator is not None:
-            s = apply_discriminator(s, discriminator, global_definitions)
-        return s
-
-    return _core_utils.walk_core_schema(schema, inner, copy=False)
+    # These are gathered for schema cleaning in pydantic-core
+    schema.setdefault('metadata', {})[CORE_SCHEMA_METADATA_DISCRIMINATOR_PLACEHOLDER_KEY] = discriminator
```
[the remark this answers is not in the document]
Comment (MarkusSintonen) [reply pydantic/_internal/_discriminated_union.py]: Done, add docstring

File: pydantic/_internal/_generate_schema.py [hunk 15]
```diff
@@ -2425,12 +2392,108 @@ def _common_field(
     }
 
 
+class CollectedInvalid(Exception):
+    pass
```
[the remark this answers is not in the document]
Comment (MarkusSintonen) [reply pydantic/_internal/_generate_schema.py]: Done

File: pydantic/_internal/_generate_schema.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (MarkusSintonen) [reply pydantic/_internal/_generate_schema.py]: Done, did the additional refactoring for these keys

File: pydantic/_internal/_generate_schema.py (hunk 4 above)
Comment (MarkusSintonen) [reply pydantic/_internal/_generate_schema.py]: Done, added some basic examples

File: tests/test_type_alias_type.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (sydney-runkle) [reply tests/test_type_alias_type.py:404]: Right, but I'm concerned that folks were dependent upon the old naming pattern for parametrized generic models in JSON schema

File: pydantic/_internal/_generate_schema.py (hunk 5 above)
Comment (sydney-runkle) [reply pydantic/_internal/_generate_schema.py:1647]: Just as a sanity check - @Viicos, I think you added that copy - does this seem alright to you?

File: tests/test_generics.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (sydney-runkle) [reply tests/test_generics.py:1843]: This is fine, thanks for explaining 👍

File: tests/test_type_alias_type.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (sydney-runkle) [reply tests/test_type_alias_type.py:404]: This PR is already so large that I'd prefer we remove this specific change. Feel free to open an issue requesting that this be changed, and we can discuss further there 👍

File: pydantic/_internal/_generate_schema.py (hunk 3 above)
Comment (sydney-runkle) [reply pydantic/_internal/_generate_schema.py:2424]: I guess I was just thinking that `self.defs._definitions.get(ref)` wasn't so bad, but I get that's not ideal (`self` here being a `GenerateSchema` instance).

Maybe a compromise, naming it `get_from_ref`? Seems like `self.defs.get_from_ref(some_ref)` sheds more light on the functionality? or `get_def_from_ref`, but that seems verbose given that we're calling it on `defs`...

**Comment by sydney-runkle** [comment]
@MarkusSintonen, awesome work with the prompt edits here. Will take another pass here, as well as on the `pydantic-core` PR as well.

File: pydantic/_internal/_generate_schema.py [hunk 16]
```diff
@@ -533,36 +529,35 @@ def _apply_discriminator_to_union(
...
+        the one just given by generate_schema. Does in place substituting to the schema by mutating it for performance
+        reasons. Use deep_copy to avoid mutating the original schema.
+
+        Example for substituting definition-refs:
+        Input:
+        {"type": "definition-ref", "schema_ref": "Xyz"}
+        Definitions:
+        {"Xyz": {"type": "string", "ref": "Xyz"}}
+        Output:
+        {"type": "string", "ref": "Xyz"}
```
Comment (sydney-runkle) [inline pydantic/_internal/_generate_schema.py:548]: Maybe this is just an oversimplification, but seems confusing that something of type `"string"` would have a ref...

File: tests/test_type_alias_type.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (MarkusSintonen) [reply tests/test_type_alias_type.py:404]: I dont think its easy to just remove as the change came from the fundamental issues with the generics handling. So it would require some further hack just for the sake of getting name like it previously was.

[the remark this answers is not in the document]
Comment (MarkusSintonen) [reply tests/test_type_alias_type.py:404]: Also looking this came from very recent PR changing type alias handling. So I dont think anyone relies on this

File: pydantic/_internal/_generate_schema.py (hunk 3 above)
Comment (MarkusSintonen) [reply pydantic/_internal/_generate_schema.py]: Done, renamed to `get_from_ref`

**Review by sydney-runkle — COMMENTED** [review]
I think something that might help make this a bit easier to review would be to cherry-pick some of the small changes over to small PRs. Specifically:

* The generic JSON schema naming changes can be moved (I think)
* `_extract_get_pydantic_json_schema` signature change
* Method additions to `_Definitions` that aren't essential for supporting the new walking pattern - like `get_def` and `reference_schema`

Again, I think these changes are valuable, but I think it would be good to reduce the scope of this PR and make tracking down issues easier.

Thanks for continuing to iterate with us here. Looking forward to adoption!

File: pydantic/_internal/_core_metadata.py [hunk 17]
```diff
@@ -41,6 +41,12 @@ class CoreMetadata(TypedDict, total=False):
     pydantic_js_updates: JsonDict
     pydantic_js_extra: JsonDict | JsonSchemaExtraCallable
 
+    pydantic_internal_tagged_union_tag: str
+    """Used internally to specify the tag used for a discriminated union."""
+
+    pydantic_internal_union_discriminator: str
+    """Used internally to specify discriminator for a discriminated union when its built deferred."""
+
```
[the remark this answers is not in the document]
Comment (MarkusSintonen) [reply pydantic/_internal/_core_metadata.py:49]: Hmm so what are the names you suggest? To drop the `pydantic_internal_xxx` prefix? But thats not consistent as others are `pydantic_js_xxx`

File: pydantic/_internal/_generate_schema.py (hunk 16 above)
Comment (MarkusSintonen) [reply pydantic/_internal/_generate_schema.py:548]: It happens with type alias usage. Eg with `TypeAdapter(TypeAliasType('A', int))` this happens exactly :D

File: pydantic/_internal/_core_metadata.py (hunk 17 above)
[the remark this answers is not in the document]
Comment (MarkusSintonen) [reply pydantic/_internal/_core_metadata.py]: > Can just add the docs up above with the others :)

Done

File: pydantic/_internal/_discriminated_union.py [hunk 18]
```diff
@@ -28,34 +24,11 @@ def __init__(self, ref: str) -> None:
...
-        discriminator = metadata.pop(CORE_SCHEMA_METADATA_DISCRIMINATOR_PLACEHOLDER_KEY, None)
-        if discriminator is not None:
-            s = apply_discriminator(s, discriminator, global_definitions)
-        return s
-
-    return _core_utils.walk_core_schema(schema, inner, copy=False)
+    """Sets the discriminator in the metadata of the schema. Used for deferred discriminator application.
+    These are gathered for schema cleaning in pydantic-core.
+    """
+    metadata: CoreMetadata = schema.setdefault('metadata', {})  # pyright: ignore[reportAssignmentType]
```
[the remark this answers is not in the document]
Comment (MarkusSintonen) [reply pydantic/_internal/_discriminated_union.py]: Done, although there is the typing issue

File: pydantic/_internal/_schema_generation_shared.py [hunk 19]
```diff
@@ -115,12 +114,25 @@ def resolve_ref_schema(self, maybe_ref_schema: core_schema.CoreSchema) -> core_s
...
+def get_existing_core_schema(obj: Any) -> core_schema.CoreSchema | None:
+    # Only use the cached value from this _exact_ class; we don't want one from a parent class
+    # This is why we check `cls.__dict__` and don't use `cls.__pydantic_core_schema__` or similar.
+    if (
+        hasattr(obj, '__dict__')
+        and (existing_schema := obj.__dict__.get('__pydantic_core_schema__')) is not None
+        and not isinstance(existing_schema, MockCoreSchema)
+    ):
+        return existing_schema
+    return None
```
[the remark this answers is not in the document]
Comment (MarkusSintonen) [reply pydantic/_internal/_schema_generation_shared.py]: Done

**Comment by MarkusSintonen** [comment]
> The generic JSON schema naming changes can be moved (I think)

What are these? I dont think this PR does any changes to JSON schemas.

> _extract_get_pydantic_json_schema signature change

Done, brought back the unused arg

> Method additions to _Definitions that aren't essential for supporting the new walking pattern - like get_def and reference_schema

Hmm but they are as it no longer does the redundant schema gathers. So we need there to create the refs and register the defs properly (and protect them from GenerateSchema side mutations).

File: pydantic/_internal/_core_metadata.py (hunk 17 above)
[the remark this answers is not in the document]
Comment (sydney-runkle) [reply pydantic/_internal/_core_metadata.py]: Sorry, to clarify, I think we should keep `pydantic_internal_XXX`, I was just referring to the second part.

Maybe:
`pydantic_internal_union_tag`
`pydantic_internal_union_discriminator`

File: pydantic/types.py [hunk 20]
```diff
@@ -3034,10 +3034,9 @@ def _convert_schema(self, original_schema: core_schema.CoreSchema) -> core_schem
             tag = None
             if isinstance(choice, tuple):
                 choice, tag = choice
-            metadata = choice.get('metadata')
+            metadata: _core_metadata.CoreMetadata | None = choice.get('metadata')  # pyright: ignore[reportAssignmentType]
```
[the remark this answers is not in the document]
Comment (sydney-runkle) [reply pydantic/types.py]: What's the issue?

**Comment by sydney-runkle** [comment]
> > The generic JSON schema naming changes can be moved (I think)

> What are these? I dont think this PR does any changes to JSON schemas. (The one thing was discussed [here](https://github.com/pydantic/pydantic/pull/10655/files#r1848842673) but it was buggy previously. Reverting that requires additional nasty hacks.)

I meant this, but I saw your recent commentary, so i think you can ignore: https://github.com/pydantic/pydantic/pull/10655#discussion_r1846844351

> Done, brought back the unused arg

Thanks - can we break that change out into a new PR? Seems like a good one.

> Hmm but they are as it no longer does the redundant schema gathers. So we need the ability to create the refs and register the defs properly (and protect them from GenerateSchema side mutations). I wouldnt want to do it manually outside the _Definitions registry without having the additional safe guards. Eg mutating the internal dicts directly feels risky and error prone.

Fair enough, good point.

**Comment by sydney-runkle** [comment]
> Hmm but they are as it no longer does the redundant schema gathers. So we need the ability to create the refs and register the defs properly (and protect them from GenerateSchema side mutations). I wouldnt want to do it manually outside the _Definitions registry without having the additional safe guards. Eg mutating the internal dicts directly feels risky and error prone

Question - can we go ahead and do this in a different PR though, and rebase this one after that is merged? Would make the diff review easier + cleaner.

**Comment by MarkusSintonen** [comment]
> Question - can we go ahead and do this in a different PR though, and rebase this one after that is merged? Would make the diff review easier + cleaner.

Well yeah, that we can do 👍 I think, maybe, lets see...

**Comment by sydney-runkle** [comment]
Hey @MarkusSintonen,

We really appreciate the work you're doing here. We're breaking out parts of this PR into smaller bits, and are definitely planning on implementing a majority of your changes before v2.11
